lib/privs.c

```diff
@@ -698,25 +708,69 @@ static int getgrouplist(const char *user, gid_t group, gid_t *groups,
 }
 #endif /* HAVE_GETGROUPLIST */
 
+/*
+ * Helper function that locates a refcounting object to use: a process-wide
+ * object or a per-pthread object.
+ */
+static struct zebra_privs_refs_t *get_privs_refs(struct zebra_privs_t *privs)
+{
+	struct zebra_privs_refs_t *temp, *refs = NULL;
+	pthread_t tid;
+
+	if (privs_per_process)
+		refs = &(privs->process_refs);
+	else {
+		/* Locate - or create - the object for the current pthread. */
+		tid = pthread_self();
+
+		STAILQ_FOREACH(temp, &(privs->thread_refs), entry) {
+			if (pthread_equal(temp->tid, tid)) {
+				refs = temp;
+				break;
+			}
+		}
+
+		/* Need to create a new refcounting object. */
+		if (refs == NULL) {
+			refs = XCALLOC(MTYPE_PRIVS,
+				       sizeof(struct zebra_privs_refs_t));
+			if (refs) {
+				refs->tid = tid;
+				STAILQ_INSERT_TAIL(&(privs->thread_refs),
+						     refs, entry);
+			}
+		}
+	}
+
+	return refs;
+}
+
 struct zebra_privs_t *_zprivs_raise(struct zebra_privs_t *privs,
 				    const char *funcname)
 {
 	int save_errno = errno;
+	struct zebra_privs_refs_t *refs;
 
 	if (!privs)
 		return NULL;
 
-	/* If we're already elevated, just return */
+	/*
+	 * Serialize 'raise' operations; particularly important for
+	 * OSes where privs are process-wide.
+	 */
 	pthread_mutex_lock(&(privs->mutex));
 	{
-		if (++(privs->refcount) == 1) {
+		/* Locate ref-counting object to use */
+		refs = get_privs_refs(privs);
+
+		if (++(refs->refcount) == 1) {
 			errno = 0;
 			if (privs->change(ZPRIVS_RAISE)) {
 				zlog_err("%s: Failed to raise privileges (%s)",
 					 funcname, safe_strerror(errno));
 			}
```

Comment:
Remind me - do we not need to make sure the refcount gets decremented if privs raising fails? Or are we simply relying on the subsequent privs drop to handle this and assume whatever error handling around privileged operations inside the elevated block from the caller handles the rest?

Reply:
yes, because of the way the `cleanup` attribute is used, I think we'll always both increment and decrement the refcount. the callers generally check for failures of their own operations inside the elevated block, and handle anything related to their own data structures.